autokeras/hypermodel/preprocessor.py

```diff
@@ -673,3 +676,233 @@ def set_config(self, config):
+            Defaults to 1000.
+    """
+
+    def __init__(self, column_types, max_columns=1000, **kwargs):
+        super().__init__(**kwargs)
+        self.column_types = column_types
+        self.max_columns = max_columns
+        self.num_columns = len(column_types)
+        self.num_rows = 0
+        self._shape = None
+        # A list of categorical column indices.
+        self.categorical_col = []
+        # A list of numerical column indices.
+        self.numerical_col = []
+        self.label_encoders = {}
+        self.value_counters = {}
+        self.categorical_categorical = {}
+        self.numerical_categorical = {}
+        self.count_frequency = {}
+        self.high_level1_col = []  # more than 32, less than 100
+        self.high_level2_col = []  # more than 100
+        self.high_level_cat_cat = {}
+        self.high_level_num_cat = {}
+
+        for index, column_type in enumerate(self.column_types):
+            if column_type == 'categorical':
+                self.categorical_col.append(index)
+            elif column_type == 'numerical':
+                self.numerical_col.append(index)
+            else:
+                raise ValueError('Unsupported column type: '
+                                 '{type}'.format(type=column_type))
+
+        for index, cat_col_index1 in enumerate(self.categorical_col):
+            self.label_encoders[cat_col_index1] = utils.LabelEncoder()
+            self.value_counters[cat_col_index1] = collections.defaultdict(
+                return_zero)
+            self.count_frequency[cat_col_index1] = {}
+            for cat_col_index2 in self.categorical_col[index + 1:]:
+                self.categorical_categorical[(
+                    cat_col_index1,
+                    cat_col_index2)] = collections.defaultdict(return_zero)
+            for num_col_index in self.numerical_col:
+                self.numerical_categorical[(
+                    num_col_index,
+                    cat_col_index1)] = collections.defaultdict(return_zero)
+
+    def update(self, x, y=None):
+        self.num_rows += 1
+        x = nest.flatten(x)[0].numpy()
+        for index in range(len(x)):
+            x[index] = x[index].decode('utf-8')
+
+        self._impute(x)
+
+        for col_index in self.categorical_col:
+            key = str(x[col_index])
+            self.label_encoders[col_index].update(key)
+            self.value_counters[col_index][key] += 1
+
+        for col_index1, col_index2 in self.categorical_categorical.keys():
+            key = (str(x[col_index1]), str(x[col_index2]))
+            self.categorical_categorical[(col_index1, col_index2)][key] += 1
+
+        for num_col_index, cat_col_index in self.numerical_categorical.keys():
+            key = str(x[cat_col_index])
+            v = x[num_col_index]
+            self.numerical_categorical[(num_col_index, cat_col_index)][key] += v
+
+    def transform(self, x, fit=False):
```

Comment:
What is the type of `x`?

Reply:
eager tensor, I expect. This is in the map function of the dataset.